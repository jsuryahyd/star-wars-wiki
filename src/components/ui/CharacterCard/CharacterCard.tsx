import React from 'react';
import {
  Box,
  Avatar,
  Heading,
  Text,
  Stack,
  LinkBox,
  LinkOverlay,
  IconButton,
  Flex,
} from '@chakra-ui/react';

import { Link as RouterLink } from '@tanstack/react-router';

interface CharacterCardProps {
  name: string;
  avatarUrl?: string;
  details: { label: string; value: string }[];
  id: string;
	renderTopRight?: (id:string) => React.ReactNode;
}

export const CharacterCard: React.FC<CharacterCardProps> = ({
  name,
  avatarUrl,
  details,
	id,
	renderTopRight
}) => {
  // Extract the character ID from the URL (assuming SWAPI URL structure)
  

  return (
    <LinkBox
      as="article"
      rounded="md"
      shadow="sm"
      transition="transform 0.15s ease-in-out"
      _hover={{ transform: 'scale(1.02)' }}
    >
      <LinkOverlay as={RouterLink} to={`/characters/${id}`}>
			{renderTopRight?.(id)}
        <Flex align="center" gap={4} p={4}>
          {/* <Avatar src={avatarUrl} name={name} size="md" /> */}
					 <Avatar.Root size="2xl" name={name} >
    <Avatar.Fallback />
    <Avatar.Image src={avatarUrl} />
  </Avatar.Root>
          <Stack spacing={1}>
            <Heading as="h3" fontSize="lg" fontWeight="semibold" mb={0}>
              {name}
            </Heading>
            {details.map((detail) => (
              <Text key={detail.label} fontSize="sm" color="gray.600">
                {/* {detail.label}:  */}
								{detail.value || 'Unknown'}
              </Text>
            ))}
          </Stack>
        </Flex>
      </LinkOverlay>
    </LinkBox>
  );
};