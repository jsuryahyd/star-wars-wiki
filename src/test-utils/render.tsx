import React from "react";
import { act, render as rtlRender } from "@testing-library/react";
import {
  RouterProvider,
  createRootRoute,
  createRouter,
  createMemoryHistory,
  createRoute,
  Outlet,
} from "@tanstack/react-router";
import { Provider } from "@/components/ui/provider";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { Toaster } from "@/components/ui/toaster";

const customRender = (
  ui: React.ReactElement,
  {
    initialRoute = "/",
    routes,
    ...options
  }: {
    initialRoute?: string;
    routes?: { path: string; component: any }[];
  } & Omit<Parameters<typeof rtlRender>[1], "wrapper"> = {}
) => {
  const rootRoute = createRootRoute({
    component: Outlet,
  });
  if (!routes?.find((r) => r.path === "/")) {
    routes?.push({
      path: "/",
      component: () => ui,
    });
  }
  const history = createMemoryHistory({
    initialEntries: initialRoute ? [initialRoute] : [],
  }); //todo: this is supposed to render the initial router immediately, avoid act(()=> router.navigate({ to: '/' }) in the test. but it doesn't work.
  const _routes = (routes || []).map((route) =>
    createRoute({ ...route, getParentRoute: () => rootRoute })
  );
  const router = createRouter({
    defaultPendingMinMs: 0,
    routeTree: rootRoute.addChildren(_routes),
    history,
  });
  // console.log(
  //   { history },
  //   { initialEntries: initialRoute ? [initialRoute] : [] },
  //   router.latestLocation
  // );
  // act(() => {
  //   // return history.push(initialRoute);
  //   router.navigate({
  //     to: "/",
  //   });
  // });
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        staleTime: 1000 * 60 * 5, // 5 minutes
        gcTime: 1000 * 60 * 10, // 10 minutes
      },
      mutations: {
        retry: false,
      },
    },
  });

  const Wrapper = ({ children }: { children: React.ReactNode }) => (
    <Provider>
      <QueryClientProvider client={queryClient}>
        {_routes.length ? (
          <RouterProvider
            router={router}
            defaultComponent={() => children}
          ></RouterProvider>
        ) : (
          children
        )}
        <Toaster />
      </QueryClientProvider>
    </Provider>
  );

  const renderReturn = rtlRender(ui, { wrapper: Wrapper, ...options });
  // await act(() =>
  //   router.navigate({

  //     to: '/',
  //   })
  // )
  // console.log(router)
  return { ...renderReturn, router };
};

export * from "@testing-library/react";
export { customRender as render, rtlRender };
