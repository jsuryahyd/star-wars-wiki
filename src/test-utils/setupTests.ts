declare const vi: typeof import('vitest')['vi'];
import '@testing-library/jest-dom';
import { server } from '../../mocks/server';

// Workarounds for JSDOM issues (if needed)
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: vi.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(), // deprecated
    removeListener: vi.fn(), // deprecated
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  })),
});

// Establish API mocking before all tests.
beforeAll(() => server.listen());
// Reset any request handlers that are declared as a part of tests (i.e. for testing one-time error scenarios)
afterEach(() => server.resetHandlers());
// Clean up after the tests are finished.
afterAll(() => server.close());

// console.log('Registering snapshot serializer');
//Dynamic classes like css-[hash] are not deterministic, so we need to replace them with a static class name for snapshot testing
//causes out of memory error
// expect.addSnapshotSerializer({
//   test: val => typeof val === 'string',
//   print: (val, serialize) =>
//     serialize((val as string).replace(/class="css-[^"]+"/g, 'class="chakra-class"')),
// });

// // Suppress Testing Library's debug HTML output on test fail -> not working
// const originalError = console.error;
// console.error = (...args) => {
//   // Filter out large HTML dumps (common pattern: contains <body or <div or <html)
//   if (
//     args.length &&
//     typeof args[0] === 'string' &&
//     /<body|<div|<html/i.test(args[0])
//   ) {
//     return;
//   }
//   originalError(...args);
// };