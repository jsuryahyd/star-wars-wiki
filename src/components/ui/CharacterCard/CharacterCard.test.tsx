import { render, screen } from "@/test-utils/render";
import { CharacterCard } from "./CharacterCard";
import userEvent from "@testing-library/user-event";

vi.mock(
  "@tanstack/react-router",
  async (importOriginal: () => Promise<any>) => {
    return {
      ...(await importOriginal()),
      Link: ({ children, to }: { children: React.ReactNode; to: string }) => (
        <a href={to}>{children}</a>
      ),
    };
  }
);
describe("CharacterCard", () => {
  it("renders correctly", () => {
    render(
      <CharacterCard
        name="Jaya Surya"
        details={[
          { label: "gender", value: "Male" },
          { label: "homePlanet", value: "Earth" },
        ]}
        id="1"
      />
    );
    expect(screen.getByRole("heading", { level: 3 })).toHaveTextContent(
      "Jaya Surya"
    );
    expect(
      screen.getByText("Male")
    ).toBeInTheDocument();
		expect(
		screen.getByText("Earth")).toBeInTheDocument();
  });

  it('matches snapshot', () => {
  	const { asFragment } = render(<CharacterCard name="Jaya Surya"
        details={[
          { label: "gender", value: "Male" },
          { label: "homePlanet", value: "Earth" },
        ]}
        id="1"/>);
  	expect(asFragment()).toMatchSnapshot();
  });

	it('renders the top right element', async () => {
		render(
			<CharacterCard
				name="Jaya Surya"
				details={[]}
				id="1"
				renderTopRight={(id)=>{
					return <span onClick={()=>{const span = document.createElement('span');span.textContent ="click success"; document.body.appendChild(span)}}>Id is {id}</span>
				}}
				/>)

		expect(screen.getByText(/Id is 1/i)).toBeInTheDocument();
		expect(screen.queryByText(/click success/i)).not.toBeInTheDocument();
		await userEvent.click(screen.getByText(/Id is 1/i));
		expect(screen.getByText(/click success/i)).toBeInTheDocument();

	})
});
