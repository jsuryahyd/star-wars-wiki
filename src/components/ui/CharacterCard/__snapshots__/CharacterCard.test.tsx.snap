// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterCard > matches snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <article
    class="chakra-linkbox css-14g7ltz"
  >
    <a
      href="/characters/1"
    >
      <div
        class="css-10zhkx9"
      >
        <div
          class="chakra-avatar__root css-pckxl"
          data-part="root"
          data-scope="avatar"
          dir="ltr"
          id="avatar:«r1»"
          name="Jaya Surya"
        >
          <span
            class="chakra-avatar__fallback css-1cuxwyn"
            data-part="fallback"
            data-scope="avatar"
            data-state="visible"
            dir="ltr"
            id="avatar:«r1»:fallback"
          >
            <svg
              class="css-1dklkmj"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 22H18V20C18 18.3431 16.6569 17 15 17H9C7.34315 17 6 18.3431 6 20V22H4V20C4 17.2386 6.23858 15 9 15H15C17.7614 15 20 17.2386 20 20V22ZM12 13C8.68629 13 6 10.3137 6 7C6 3.68629 8.68629 1 12 1C15.3137 1 18 3.68629 18 7C18 10.3137 15.3137 13 12 13ZM12 11C14.2091 11 16 9.20914 16 7C16 4.79086 14.2091 3 12 3C9.79086 3 8 4.79086 8 7C8 9.20914 9.79086 11 12 11Z"
              />
            </svg>
          </span>
          <img
            class="chakra-avatar__image css-1odw0y2"
            data-part="image"
            data-scope="avatar"
            data-state="hidden"
            dir="ltr"
            draggable="false"
            hidden=""
            id="avatar:«r1»:image"
            referrerpolicy="no-referrer"
          />
        </div>
        <div
          class="chakra-stack css-8g8ihq"
          spacing="1"
        >
          <h3
            class="chakra-heading css-9nm8pf"
          >
            Jaya Surya
          </h3>
          <p
            class="css-v0otzr"
          >
            Male
          </p>
          <p
            class="css-v0otzr"
          >
            Earth
          </p>
        </div>
      </div>
    </a>
  </article>
</DocumentFragment>
`;
